### ISSUE TITLE
multiload: permission denied on FUSE filesystems (gvfs mountentries)

### ISSUE THREAD

**Issue by arumbold** [mate-desktop/mate-applets#171]
Enabling Harddisk monitor causes the journal to get flooded with messages like

```
Dec 22 12:00:36 aaron-desktop org.mate.panel.applet.MultiLoadAppletFactory[2483]: glibtop(c=5008): [WARNING] statvfs '/run/user/42/gvfs' failed: Permission denied
```

Could an exception be added to GetDiskLoad to prevent the method from trying to poll fsusage for FUSE mounts owned by other users?
## GetDiskLoad()

```
multiload/linux-proc.c
GetDiskLoad()
```

```
mountentries = glibtop_get_mountlist (&mountlist, FALSE);
for (i = 0; i < mountlist.number; i++)
{
    glibtop_fsusage fsusage;

    if (strcmp(mountentries[i].type, "smbfs") == 0
        || strcmp(mountentries[i].type, "nfs") == 0
        || strcmp(mountentries[i].type, "cifs") == 0)
        continue;

    glibtop_get_fsusage(&fsusage, mountentries[i].mountdir);
    read += fsusage.read; write += fsusage.write;
}
```
##

**Comment by raveit65** [comment]
Well, you hit a old systemd bug, read https://bugzilla.redhat.com/show_bug.cgi?id=753882
... and upream won't fix it.
Correct permissions for /run/user/42/gvfs and all is fine.

**Comment by arumbold** [comment]
Thank you raveit.

I've had look through the referenced bug but can't work out what permissions I should be applying to this file, or how they should be added?

uid 42 is the gdm user, not myself but nobody (not even root) seems to have access to that directory.

```
[root@aaron-desktop ~] # ls -al /run/user/42/gvfs
ls: cannot access /run/user/42/gvfs: Permission denied
```
```
[root@aaron-desktop ~] # ls -alZ /run/user/42/
ls: cannot access /run/user/42/gvfs: Permission denied
total 0
drwx------. 8 gdm  gdm  system_u:object_r:user_tmp_t:s0    160 Dec 22 09:56 .
drwxr-xr-x. 4 root root system_u:object_r:user_tmp_t:s0     80 Dec 22 09:56 ..
drwx------. 2 gdm  gdm  system_u:object_r:config_home_t:s0  60 Dec 22 09:55 dconf
drwx--x--x. 2 gdm  gdm  system_u:object_r:user_tmp_t:s0     60 Dec 22 09:55 gdm
drwx------. 3 gdm  gdm  system_u:object_r:user_tmp_t:s0     60 Dec 22 09:56 gnome-shell
d?????????? ? ?    ?                                     0   ?            ? gvfs
drwx------. 2 gdm  gdm  system_u:object_r:user_tmp_t:s0     80 Dec 22 09:56 pulse
drwxr-xr-x. 2 gdm  gdm  system_u:object_r:user_tmp_t:s0     80 Dec 22 09:55 systemd
[root@aaron-desktop ~] # 
```

**Comment by arumbold** [comment]
I went back to re-read the original redhat bug.

That seems to be something different to the issue that I'm experiencing?
my uid is 1000:
```
[aaron@aaron-desktop ~] $ echo $XDG_RUNTIME_DIR
/run/user/1000
```
the journal log contains uid 42
```
Dec 22 17:02:31 aaron-desktop org.mate.panel.applet.MultiLoadAppletFactory[2483]: glibtop(c=2949): [WARNING] statvfs '/run/user/42/gvfs' failed: Permission denied
Dec 22 17:02:32 aaron-desktop org.mate.panel.applet.MultiLoadAppletFactory[2483]: glibtop(c=2949): [WARNING] statvfs '/run/user/42/gvfs' failed: Permission denied
Dec 22 17:02:32 aaron-desktop org.mate.panel.applet.MultiLoadAppletFactory[2483]: glibtop(c=2949): [WARNING] statvfs '/run/user/42/gvfs' failed: Permission denied
Dec 22 17:02:33 aaron-desktop org.mate.panel.applet.MultiLoadAppletFactory[2483]: glibtop(c=2949): [WARNING] statvfs '/run/user/42/gvfs' failed: Permission denied
Dec 22 17:02:33 aaron-desktop org.mate.panel.applet.MultiLoadAppletFactory[2483]: glibtop(c=2949): [WARNING] statvfs '/run/user/42/gvfs' failed: Permission denied
```

in addition to the mounts you would expect, `mount -v ` contains the following:
```
sunrpc on /var/lib/nfs/rpc_pipefs type rpc_pipefs (rw,relatime)
tmpfs on /run/user/42 type tmpfs (rw,nosuid,nodev,relatime,seclabel,size=1634312k,mode=700,uid=42,gid=42)
fusectl on /sys/fs/fuse/connections type fusectl (rw,relatime)
gvfsd-fuse on /run/user/42/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=42,group_id=42)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,seclabel,size=1634312k,mode=700,uid=1000,gid=1000)
gvfsd-fuse on /run/user/1000/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,relatime)
gvfsd-fuse on /home/aaron/.gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
```
You can see some of the fuse connections reporting a user_id of 42 (__gdm__)

For reference, I am running Fedora 23 Workstation if that makes any difference.

**Comment by raveit65** [comment]
Good, i'm using fedora too but not gdm, i'm using lightdm.
Well, i can try to reproduce it.

**Comment by raveit65** [comment]
Confirmed with f22/23 VMs.
But why use gdm a system uid?
Lightdm use 997 uid and i see same output with 'mount -v'
```
tmpfs on /run/user/997 type tmpfs (rw,nosuid,nodev,relatime,size=204924k,mode=700,uid=987,gid=981)
```
I see not really a different here, exept that lightdm does not use a system account.

This messages will spam your journal, using 'journalctl -b0 -f --no-tail' shows me that those messages don't stop, bad for a box which runs all the day.
I suggest to use lightdm for the moment as workaround.
....i hate gnome software...

**Comment by rabin-io** [comment]
I encounter this error when running docker containers, I opened a bug report on [bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1355867) before I found this issue in here.

https://bugzilla.redhat.com/show_bug.cgi?id=1355867

**Comment by rabin-io** [comment]
With Fedora 27 this is still an issue, 
one thing to mention about my original problem with docker is that in journald i see this messages over and over, 
    org.mate.panel.applet.MultiLoadAppletFactory[2612]: glibtop(c=2992): [WARNING] statvfs '/var/lib/docker/devicemapper/mnt/ce7960b5dfc3bd0759869f827cd503b5b43416cc677ea7' failed: Permission denied

but this is not the full path, the full path is, 
    /var/lib/docker/devicemapper/mnt/ce7960b5dfc3bd0759869f827cd503b5b43416cc677ea743e807d4dc35780837

seems like the path is trimmed at 80 characters

**Comment by raveit65** [comment]
Probably the same issue
https://github.com/mate-desktop/mate-applets/issues/230

**Comment by jedie** [comment]
> I encounter this error when running docker containers

Same here with UbuntuMate 17.10
```
$ uname -a
Linux 4.13.0-21-generic #24-Ubuntu SMP Mon Dec 18 17:29:16 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
$ docker version
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:11:14 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:09:47 2017
  OS/Arch:	linux/amd64
  Experimental:	false
```

**Comment by Sanva** [comment]
The thing is, if this is the same problem of #230 I have a suggestion for a fix there [1] —but I simply don't know if ignoring the `nsfs` *mountentries* is the correct thing to do... nor why there are other *mountentry*-types being ignored in https://github.com/Sanva/mate-applets/blob/master/multiload/linux-proc.c#L124 — at that time I wasn't able to find resources about this.

[1] https://github.com/mate-desktop/mate-applets/issues/230#issuecomment-285940631

**Comment by marosg42** [comment]
@Sanva I think rationale behind ignoring nfs, smb and cifs is that those are usually network mounts, so they will not provide you any useful disk load stats and they can be ignored.

**Comment by tofurky** [comment]
i've encountered the same/similar issue. it seemed to get triggered by things being mounted/unmounted during an invocation of pmbootstrap (an alpine linux embedded build utility). afaik these weren't network or fuse mounts but instead bind mounts done by root via sudo. my guess is that at least one cause of the log spam is the mount being detected but then disappearing, rather than not having permissions. the applet continues to attempt gathering stats on the non-existent mount without refreshing /proc/mounts:
```
user.warning<12> Feb 24 00:00:00 aquos org.mate.panel.applet.MultiLoadA glibtop(c=1673): [WARNING] statvfs '/home/matt/devel/ouya/pmbootstrap-work/chroot_native/mnt/pmbootstrap-abuild-con' failed: No such file or directory
user.warning<12> Feb 24 00:00:00 aquos org.mate.panel.applet.MultiLoadA glibtop(c=1673): [WARNING] statvfs '/home/matt/devel/ouya/pmbootstrap-work/chroot_buildroot_armhf/mnt/pmbootstrap-c' failed: No such file or directory
user.warning<12> Feb 24 00:00:00 aquos org.mate.panel.applet.MultiLoadA glibtop(c=1673): [WARNING] statvfs '/home/matt/devel/ouya/pmbootstrap-work/chroot_buildroot_armhf/var/cache/distfil' failed: No such file or directory
user.warning<12> Feb 24 00:00:00 aquos org.mate.panel.applet.MultiLoadA glibtop(c=1673): [WARNING] statvfs '/home/matt/devel/ouya/pmbootstrap-work/chroot_buildroot_armhf/mnt/pmbootstrap-g' failed: No such file or directory

```

**Comment by jodumont** [comment]
it also affect Debian9 as mentionned here]([https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=855037)

**Comment by lukefromdc** [comment]
The only non-local mounts I can set up are sshfs or loopback filesystems. Will those do for testing this?

**Comment by rbuj** [comment]
`glibtop_get_fsusage` should only get fsusage for block type devices such as hard drives (e. g. /dev/sda device), excluding external devices (e. g. /media/cdrom mount point)
https://github.com/mate-desktop/mate-applets/blob/c24ac666e9d77f9be866c8603b22d11e534dde7e/multiload/linux-proc.c#L138

**Comment by tofurky** [comment]
@raveit65 thanks, although this will not fix the case of a filesystem having existed, then gone away, i think. i.e. https://github.com/mate-desktop/mate-applets/issues/171#issuecomment-389587125

**Comment by raveit65** [comment]
Are you sure this should be re-opened?

**Comment by tofurky** [comment]
i'll have to build from master and see if it's still an issue. it's really a different bug with similar symptoms.

**Comment by rbuj** [comment]
all_fs is FALSE in glibtop_get_mountlist call:
https://github.com/mate-desktop/mate-applets/blob/736ba510c9d5c55ca5a3a59693ce25fba010ad03/multiload/linux-proc.c#L119

Ignored fs if all_fs is FALSE:
```
$ cat /proc/filesystems | grep nodev
nodev	sysfs
nodev	rootfs
nodev	ramfs
nodev	bdev
nodev	proc
nodev	cpuset
nodev	cgroup
nodev	cgroup2
nodev	tmpfs
nodev	devtmpfs
nodev	configfs
nodev	debugfs
nodev	tracefs
nodev	securityfs
nodev	sockfs
nodev	dax
nodev	bpf
nodev	pipefs
nodev	hugetlbfs
nodev	devpts
nodev	autofs
nodev	mqueue
nodev	selinuxfs
nodev	pstore
nodev	rpc_pipefs
nodev	fuse
nodev	fusectl
```
https://github.com/GNOME/libgtop/blob/abccaf488a929de1e95e6a748485575dec52c998/sysdeps/linux/mountlist.c#L99-L102

**Comment by tofurky** [comment]
it turns out the issues i was having were due to the mounts having over 79 characters in the path. there is a hard-coded limit in libgtop here: https://github.com/GNOME/libgtop/blob/abccaf488a929de1e95e6a748485575dec52c998/glibtop.h#L63-L65

unfortunately, after installing the recompiled modified libgtop (bumping GLIBTOP_MOUNTENTRY_LEN from 79 -> 4096) i ran into issues with the precompiled mate-multiload-applet in ubuntu due to it having been compiled with a libgtop.h which specified 79 chars. after rebuilding mate-applets against the modified libgtop headers, the issue went away.

so, in other words, it was an entirely different issue :) thanks
